src/Compilers/Core/CodeAnalysisTest/ObjectSerializationTests.cs

```diff
@@ -5,16 +5,373 @@
+            }
+
+            Func<ObjectReader, object> IObjectReadable.GetReader() => (r) => new TypeWithManyMembers<T>(r);
+
+            public override int GetHashCode()
+            {
+                return _members.Length;
+            }
+
+            public override Boolean Equals(Object obj)
+            {
+                return Equals(obj as TypeWithManyMembers<T>);
+            }
+
+            public bool Equals(TypeWithManyMembers<T> other)
+            {
+                if (other == null)
+                {
+                    return false;
+                }
+
+                if (_members.Length != other._members.Length)
+                {
+                    return false;
+                }
+
+                return Equalish(_members, other._members);
+            }
+        }
+
+        private void TestRoundTripMember<T>(T value)
+        {
+            TestRoundTripValue(new TypeWithOneMember<T>(value));
+        }
+
+        private void TestRoundTripMembers<T, S>(T value1, S value2)
+        {
+            TestRoundTripValue(new TypeWithTwoMembers<T, S>(value1, value2));
+        }
+
+        private void TestRoundTripMembers<T>(params T[] values)
+        {
+            TestRoundTripValue(new TypeWithManyMembers<T>(values));
+        }
+
+        [Fact]
+        public void TestValueInt32()
+        {
+            TestRoundTripValue(123);
+        }
+
+        [Fact]
+        public void TestMemberInt32()
+        {
+            TestRoundTripMember(123);
+        }
+
+        [Fact]
+        public void TestMemberIntString()
+        {
+            TestRoundTripMembers(123, "Hello");
+        }
+
+        [Fact]
+        public void TestManyMembersInt32()
+        {
+            TestRoundTripMembers(Enumerable.Range(0, 1000).ToArray());
+        }
+
+        [Fact]
+        public void TestSmallArrayMember()
+        {
+            TestRoundTripMember(Enumerable.Range(0, 3).ToArray());
+        }
+
+        [Fact]
+        public void TestEmptyArrayMember()
+        {
+            TestRoundTripMember(new int[] { });
+        }
+
         [Fact]
-        public void TestRoundTripPrimitiveArray()
+        public void TestNullArrayMember()
+        {
+            TestRoundTripMember<int[]>(null);
+        }
+
+        [Fact]
+        public void TestLargeArrayMember()
+        {
+            TestRoundTripMember(Enumerable.Range(0, 1000).ToArray());
+        }
+
+        [Fact]
+        public void TestEnumMember()
+        {
+            TestRoundTripMember(EByte.Value);
+        }
+
+        [Fact]
+        public void TestPrimitiveArrayValues()
+        {
+            TestRoundTrip(w => TestWritingPrimitiveArrays(w), r => TestReadingPrimitiveArrays(r));
+        }
+
+
+        private void TestRoundTripCompressedUint(uint value)
+        {
+            TestRoundTrip(value, (w, v) => ((StreamObjectWriter)w).WriteCompressedUInt(v), r => ((StreamObjectReader)r).ReadCompressedUInt());
+        }
+
+        [Fact]
```

Comment:
Facts need some sort of Trait that categorizes what code area this is.  A trait for serialization would be fine.